consensus/src/main/java/org/apache/iotdb/consensus/ratis/RatisConsensus.java

```diff
@@ -21,70 +21,452 @@
+   */
+  private void syncGroupInfoAndRebuildClient(ConsensusGroupId groupId)
+      throws ConsensusGroupNotExistException {
+    RaftGroupId raftGroupId = Utils.toRatisGroupId(groupId);
+    RaftClient current = clientMap.get(raftGroupId);
+    try {
+      GroupInfoReply reply = current.getGroupManagementApi(myself.getId()).info(raftGroupId);
+
+      if (!reply.isSuccess()) {
+        throw new ConsensusGroupNotExistException(groupId);
+      }
+
+      raftGroupMap.put(raftGroupId, reply.getGroup());
+      buildClientAndCache(raftGroupMap.get(raftGroupId));
+    } catch (IOException e) {
+      throw new ConsensusGroupNotExistException(groupId);
+    }
+  }
+
+  private RaftClientReply sendReconfiguration(RaftGroupId raftGroupId, List<RaftPeer> peers)
+      throws RatisRequestFailedException {
+    RaftClient client = clientMap.get(raftGroupId);
+    // notify the group leader of configuration change
+    RaftClientReply reply = null;
+    try {
+      reply = client.admin().setConfiguration(peers);
+    } catch (IOException e) {
+      throw new RatisRequestFailedException();
+    }
+    return reply;
+  }
 
   @Override
-  public void start() {}
+  public void start() {
+    // TODO awkward, should we add IOException to method signature?
+    try {
+      server.start();
+    } catch (IOException ignored) {
+      try {
+        server.close();
+      } catch (IOException ignored1) {
+      }
+    }
+  }
 
   @Override
-  public void stop() {}
+  public void stop() {
+    try {
+      server.close();
+    } catch (IOException ignored) {
+    }
+  }
 
   @Override
   public ConsensusWriteResponse write(
       ConsensusGroupId groupId, IConsensusRequest IConsensusRequest) {
-    return ConsensusWriteResponse.newBuilder().build();
+
+    RaftClient client = clientMap.get(Utils.toRatisGroupId(groupId));
+    TSStatus writeResult = null;
+    try {
+      ByteBufferConsensusRequest request = (ByteBufferConsensusRequest) IConsensusRequest;
+      Message message = Message.valueOf(ByteString.copyFrom(request.getContent()));
+      RaftClientReply reply = client.io().send(message);
+      writeResult = Utils.deserializeFrom(reply.getMessage().getContent().asReadOnlyByteBuffer());
+    } catch (IOException | TException e) {
+      return ConsensusWriteResponse.newBuilder()
+          .setException(new RatisRequestFailedException())
+          .build();
+    }
+    return ConsensusWriteResponse.newBuilder().setStatus(writeResult).build();
   }
 
+  /** Read directly from LOCAL COPY notice: May read stale data (not linearizable) */
   @Override
-  public ConsensusReadResponse read(ConsensusGroupId groupId, IConsensusRequest IConsensusRequest) {
-    return ConsensusReadResponse.newBuilder().build();
+  public synchronized ConsensusReadResponse read(
+      ConsensusGroupId groupId, IConsensusRequest IConsensusRequest) {
+
+    RaftClientReply reply = null;
+    try {
+      assert IConsensusRequest instanceof ByteBufferConsensusRequest;
```

Comment:
why assert this?

Reply:
RatisConsensus requires user Request implements the ByteBufferRequest, otherwise there's no way for RatisConsensus to serialize and deserialize a common IConsensusRequest to ByteString

Reply:
Then, use if statement and throw exceptions would better~